core/src/main/scala/kafka/coordinator/TransactionCoordinator.scala

```diff
@@ -0,0 +1,139 @@
+  *
+  * Unless required by applicable law or agreed to in writing, software
+  * distributed under the License is distributed on an "AS IS" BASIS,
+  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  * See the License for the specific language governing permissions and
+  * limitations under the License.
+  */
+package kafka.coordinator
+
+import java.util.concurrent.atomic.AtomicBoolean
+
+import kafka.server.{KafkaConfig, ReplicaManager}
+import kafka.utils.{Logging, Pool, ZkUtils}
+import org.apache.kafka.common.protocol.Errors
+import org.apache.kafka.common.utils.Time
+
+/**
+  * Transaction coordinator handles message transactions sent by producers and communicate with brokers
+  * to update ongoing transaction's status.
+  *
+  * Each Kafka server instantiates a transaction coordinator which is responsible for a set of
+  * producers. Producers with specific transactional ids are assigned to their corresponding coordinators;
+  * Producers with no specific transactional id may talk to a random broker as their coordinators.
+  */
+object TransactionCoordinator {
+
+  def apply(config: KafkaConfig, zkUtils: ZkUtils, time: Time): TransactionCoordinator = {
+
+    val pIDManager = new PidManager(config.brokerId, zkUtils)
+    val logManager = new TransactionLogManager(config.brokerId, zkUtils)
+    new TransactionCoordinator(config.brokerId, pIDManager, logManager)
+  }
+}
+
+class TransactionCoordinator(val brokerId: Int,
+                             val pIDManager: PidManager,
+                             val logManager: TransactionLogManager) extends Logging {
+
+  this.logIdent = "[Transaction Coordinator " + brokerId + "]: "
+
+  type InitPidCallback = InitPidResult => Unit
+
+  /* Active flag of the coordinator */
+  private val isActive = new AtomicBoolean(false)
+
+  /* TransactionalId to pid metadata map cache */
+  private val pIDMetadataCache = new Pool[String, PidMetadata]
+
+  def handleInitPid(transactionalId: String,
+                    responseCallback: InitPidCallback): Unit = {
+    if (transactionalId == null || transactionalId.isEmpty) {
+      // if the transactional id is not specified, then always blindly accept the request
+      // and return a new pid from the pid manager
+      val pID: Long = pIDManager.getNewPid()
+
+      responseCallback(InitPidResult(pID, epoch = 0, Errors.NONE))
+    } else if(!logManager.isCoordinatorFor(transactionalId)) {
+      // check if it is the assigned coordinator for the transactional id
+      responseCallback(initPidError(Errors.NOT_COORDINATOR_FOR_GROUP))
+    } else {
+      // only try to get a new pid and update the cache if the transactional id is unknown
+      getPidMetadata(transactionalId) match {
+        case None =>
+          val pID: Long = pIDManager.getNewPid()
+          val metadata = addPidMetadata(transactionalId, new PidMetadata(pID))
+
+          responseCallback(initPidMetadata(metadata))
```

Comment:
The case I'm imagining is two processes with the same transactionalId both send InitPidRequest at the same time. Both hit the `None` case because there was previously no PID mapped to the transactionalId. Only one of the processes will succeed in creating the associated PID metadata (because of the `putIfNotExists`), but both will get a successful response with the new epoch. Isn't that a problem?